java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql

```diff
@@ -0,0 +1,107 @@
+/**
+ * @id java/insecure-smtp-ssl
+ * @name Insecure JavaMail SSL Configuration
+ * @description Java application configured to use authenticated mail session over SSL does not validate the SSL certificate to properly ensure that it is actually associated with that host.
+ * @kind problem
+ * @tags security
+ *       external/cwe-297
+ */
+
+import java
+
+/**
+ * The method to set Java properties
+ */
+class SetPropertyMethod extends Method {
+  SetPropertyMethod() {
+    this.hasName("setProperty") and
+    this.getDeclaringType().hasQualifiedName("java.util", "Properties")
+    or
+    this.hasName("put") and
+    this.getDeclaringType().getASourceSupertype*().hasQualifiedName("java.util", "Dictionary")
+  }
+}
+
+/**
+ * The insecure way to set Java properties in mail sessions.
+ * 1. Set the mail.smtp.auth property to provide the SMTP Transport with a username and password when connecting to the SMTP server or
+ *    set the mail.smtp.ssl.socketFactory/mail.smtp.ssl.socketFactory.class property to create an SMTP SSL socket.
+ * 2. No mail.smtp.ssl.checkserveridentity property is enabled.
+ */
+predicate isInsecureMailPropertyConfig(VarAccess propertiesVarAccess) {
+  exists(MethodAccess ma |
+    ma.getMethod() instanceof SetPropertyMethod and
+    ma.getQualifier() = propertiesVarAccess.getVariable().getAnAccess() and
+    (
+      getStringValue(ma.getArgument(0)).indexOf(".auth") != -1 and //mail.smtp.auth
+      getStringValue(ma.getArgument(1)) = "true"
+      or
+      getStringValue(ma.getArgument(0)).indexOf(".socketFactory") != -1 //mail.smtp.socketFactory or mail.smtp.socketFactory.class
+    )
+  ) and
+  not exists(MethodAccess ma |
+    ma.getMethod() instanceof SetPropertyMethod and
+    ma.getQualifier() = propertiesVarAccess.getVariable().getAnAccess() and
+    (
+      getStringValue(ma.getArgument(0)).indexOf(".ssl.checkserveridentity") != -1 and //mail.smtp.ssl.checkserveridentity
+      getStringValue(ma.getArgument(1)) = "true"
+    )
+  )
+}
+
+/**
+ * Helper method to get string value of an argument
+ */
+string getStringValue(Expr expr) {
+  result = expr.(StringLiteral).getRepresentedString()
+  or
+  exists(Variable v | expr = v.getAnAccess() |
+    result = getStringValue(v.getInitializer().(CompileTimeConstantExpr))
```

Comment:
This gets the value from the initializer even if the variable is subsequently modified.  I assume you meant to restrict this to the value of `CompileTimeConstantExpr`s?  In that case, both this case and the base case above can be replaced with `result = expr.(CompileTimeConstantExpr).getStringValue()`.

Reply:
Thanks @aschackmull . I've incorporated the change to simplify the query.